userops_ext.go

```diff
@@ -161,6 +193,84 @@ func (op *UserOperation) Validate() (UserOpSolvedStatus, error) {
 	return SolvedUserOp, nil
 }
 
+// validateCrossChainOp validates a cross-chain operation
+func (op *UserOperation) validateCrossChainOp() (UserOpSolvedStatus, error) {
+	dataBytes := op.CallData
+
+	if len(dataBytes) < OpTypeLength+CallDataLengthSize {
+		return UnknownUserOp, ErrMissingCrossChainData
+	}
+
+	intentLength := int(binary.BigEndian.Uint16(dataBytes[OpTypeLength : OpTypeLength+CallDataLengthSize]))
+	expectedLength := OpTypeLength + CallDataLengthSize + intentLength + HashListLengthSize
+
+	if len(dataBytes) < expectedLength {
+		return UnknownUserOp, ErrMissingCrossChainData
+	}
+
+	hashListLength := int(dataBytes[OpTypeLength+CallDataLengthSize+intentLength])
+	if hashListLength < MinOpCount || hashListLength > MaxOpCount {
+		return UnknownUserOp, ErrInvalidHashListLength
+	}
+
+	expectedLength += hashListLength * HashListEntrySize
+
+	if len(dataBytes) != expectedLength {
+		return UnknownUserOp, ErrInvalidHashListEntry
+	}
+
+	// Validate hash list entries
+	hashListStart := OpTypeLength + CallDataLengthSize + intentLength + HashListLengthSize
+	for i := 0; i < hashListLength; i++ {
+		entryStart := hashListStart + i*HashListEntrySize
+
+		// Validate placeholder
+		placeholder := binary.BigEndian.Uint16(dataBytes[entryStart : entryStart+PlaceholderSize])
+		if placeholder != CrossChainMarker {
+			return UnknownUserOp, ErrInvalidHashListEntry
+		}
+
+		// Validate operation hash
+		operationHash := dataBytes[entryStart+PlaceholderSize : entryStart+HashListEntrySize]
+		if !validateOperationHash(operationHash) {
+			return UnknownUserOp, ErrInvalidHashListEntry
+		}
+	}
+
+	if op.HasSignature() {
+		return SolvedUserOp, nil
+	}
+
+	return UnsolvedUserOp, nil
+}
+
+// validateOperationHash checks if the provided operation hash is valid
+func validateOperationHash(hash []byte) bool {
+	// Ensure the hash is exactly 32 bytes long
+	if len(hash) != OperationHashSize {
+		return false
+	}
+
+	// Check if the hash is not all zeros (a simple validity check)
+	isAllZeros := true
+	for _, b := range hash {
+		if b != 0 {
+			isAllZeros = false
+			break
+		}
+	}
+
+	return !isAllZeros
+}
+
+// isCrossChainOperation checks if the UserOperation is a cross-chain operation
+func (op *UserOperation) isCrossChainOperation() (bool, error) {
+	if len(op.CallData) < OpTypeLength {
+		return false, nil
+	}
+	return binary.BigEndian.Uint16(op.CallData[:OpTypeLength]) == CrossChainMarker, nil
```

Comment:
It works only for unsolved ops; the x-payload would be appended to the signature payload in solved ops.